### ISSUE TITLE
IndexOutOfRangeException on Xamarin Android

### ISSUE THREAD

**Issue by r-work** [dotnet/efcore#22665]
This happens with the first request, this worked in EF 3.1.8

### To Reproduce

You can use the full app repro [here](https://github.com/r-work/XamrainEFCoreTest) or use the code below:

``` C#
public static class Repro
    {
        public static void Test()
        {
            using (var cntxt = new MyDbContext())
            {
                cntxt.Set<SubModel>(); //<- throws here
            }
        }
    }

    public class MyDbContext : DbContext
    {
        public DbSet<SubModel> SubModels { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlite("Data Source = :memory:");
        }
    }

    public class SubModel : TestModel
    {
        public string B { get; set; }
        public string C { get; set; }
        public string D { get; set; }
        public string E { get; set; }
    }

    public abstract class TestModel
    {
        [Key]
        public int Id { get; set; }
        public string A { get; set; }
    }
```

```  at System.Attribute.InternalParamGetCustomAttributes (System.Reflection.ParameterInfo parameter, System.Type attributeType, System.Boolean inherit) [0x0005b] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/attribute.cs:62 
  at System.Attribute.GetCustomAttributes (System.Reflection.ParameterInfo element, System.Type attributeType, System.Boolean inherit) [0x0008b] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/attribute.cs:686 
  at System.Attribute.GetCustomAttribute (System.Reflection.ParameterInfo element, System.Type attributeType, System.Boolean inherit) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/attribute.cs:754 
  at System.Attribute.GetCustomAttribute (System.Reflection.ParameterInfo element, System.Type attributeType) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/attribute.cs:747 
  at System.Reflection.CustomAttributeExtensions.GetCustomAttribute (System.Reflection.ParameterInfo element, System.Type attributeType) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/reflection/CustomAttributeExtensions.cs:22 
  at System.Reflection.CustomAttributeExtensions.GetCustomAttribute[T] (System.Reflection.ParameterInfo element) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/reflection/CustomAttributeExtensions.cs:39 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.NonNullableConventionBase.IsNonNullableReferenceType (Microsoft.EntityFrameworkCore.Metadata.Builders.IConventionModelBuilder modelBuilder, System.Reflection.MemberInfo memberInfo) [0x00053] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.NonNullableReferencePropertyConvention.Process (Microsoft.EntityFrameworkCore.Metadata.Builders.IConventionPropertyBuilder propertyBuilder) [0x00016] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.NonNullableReferencePropertyConvention.ProcessPropertyAdded (Microsoft.EntityFrameworkCore.Metadata.Builders.IConventionPropertyBuilder propertyBuilder, Microsoft.EntityFrameworkCore.Metadata.Conventions.IConventionContext`1[TMetadata] context) [0x00000] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher+ImmediateConventionScope.OnPropertyAdded (Microsoft.EntityFrameworkCore.Metadata.Builders.IConventionPropertyBuilder propertyBuilder) [0x00057] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher+OnPropertyAddedNode.Run (Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher dispatcher) [0x0000c] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher+DelayedConventionScope.Run (Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher dispatcher) [0x0001e] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher+ConventionBatch.Run () [0x000d5] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher+ConventionBatch.Dispose () [0x0001d] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher+ImmediateConventionScope.OnModelInitialized (Microsoft.EntityFrameworkCore.Metadata.Builders.IConventionModelBuilder modelBuilder) [0x0006f] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelInitialized (Microsoft.EntityFrameworkCore.Metadata.Builders.IConventionModelBuilder modelBuilder) [0x00000] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Metadata.Internal.Model..ctor (Microsoft.EntityFrameworkCore.Metadata.Conventions.ConventionSet conventions, Microsoft.EntityFrameworkCore.Infrastructure.ModelDependencies modelDependencies) [0x00096] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.ModelBuilder..ctor (Microsoft.EntityFrameworkCore.Metadata.Conventions.ConventionSet conventions, Microsoft.EntityFrameworkCore.Infrastructure.ModelDependencies modelDependencies, System.Boolean _) [0x00012] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.ModelBuilder..ctor (Microsoft.EntityFrameworkCore.Metadata.Conventions.ConventionSet conventions, Microsoft.EntityFrameworkCore.Infrastructure.ModelDependencies modelDependencies) [0x00000] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel (Microsoft.EntityFrameworkCore.DbContext context, Microsoft.EntityFrameworkCore.Metadata.Conventions.Infrastructure.IConventionSetBuilder conventionSetBuilder, Microsoft.EntityFrameworkCore.Infrastructure.ModelDependencies modelDependencies) [0x00012] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel (Microsoft.EntityFrameworkCore.DbContext context, Microsoft.EntityFrameworkCore.Metadata.Conventions.Infrastructure.IConventionSetBuilder conventionSetBuilder, Microsoft.EntityFrameworkCore.Infrastructure.ModelDependencies modelDependencies) [0x00046] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel () [0x00044] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model () [0x00020] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder+<>c.<TryAddCoreServices>b__7_3 (System.IServiceProvider p) [0x00006] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory (Microsoft.Extensions.DependencyInjection.ServiceLookup.FactoryCallSite factoryCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context) [0x0000d] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[TArgument,TResult].VisitCallSiteMain (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, TArgument argument) [0x00033] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngine, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverLock lockType) [0x00059] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite singletonCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context) [0x00029] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[TArgument,TResult].VisitCallSite (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, TArgument argument) [0x00057] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor (Microsoft.Extensions.DependencyInjection.ServiceLookup.ConstructorCallSite constructorCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context) [0x0002c] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[TArgument,TResult].VisitCallSiteMain (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, TArgument argument) [0x0004f] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngine, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverLock lockType) [0x00059] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite singletonCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context) [0x00029] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[TArgument,TResult].VisitCallSite (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, TArgument argument) [0x00057] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope) [0x00012] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine+<>c__DisplayClass1_0.<RealizeService>b__0 (Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope) [0x0003e] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService (System.Type serviceType, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngineScope) [0x0003e] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService (System.Type serviceType) [0x00013] in <33d3e218d40a4945a4e2f89375128b97>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService (System.IServiceProvider provider, System.Type serviceType) [0x00034] in <fa36ec453eb74eebbf8606bbb413059f>:0 
  at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T] (System.IServiceProvider provider) [0x0000e] in <fa36ec453eb74eebbf8606bbb413059f>:0 
  at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies () [0x00017] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider () [0x000bf] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies () [0x00010] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at Microsoft.EntityFrameworkCore.DbContext.Set[TEntity] () [0x00000] in <cf38e32a92d340e3abe37a488db956b3>:0 
  at XamrainEFCoreTest.Repro.Test () [0x00008] in C:\Work\XamrainEFCoreTest\XamrainEFCoreTest\XamrainEFCoreTest\Repro.cs:15 
  at XamrainEFCoreTest.App.OnStart () [0x00001] in C:\Work\XamrainEFCoreTest\XamrainEFCoreTest\XamrainEFCoreTest\App.xaml.cs:22 
  at Xamarin.Forms.Application.SendStart () [0x00000] in D:\a\1\s\Xamarin.Forms.Core\Application.cs:346 
  at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.OnStateChanged () [0x0001b] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:528 
  at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.OnStart () [0x00046] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FormsAppCompatActivity.cs:433 
  at Android.App.Activity.n_OnStart (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <0623d8432cef4751a77985c521d75c2f>:0 
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.14(intptr,intptr)
```

### Additional context

Microsoft.Data.Sqlite version: 5.0.0-rc.1.20451.13
Target framework: .NET Standard 2.1
Operating system: Android

**Comment by ajcvickers** [comment]
@roji @JeremyLikness Original issue here was #21430, which we closed as a duplicate of #18793, which itself was closed as external. Can we check where we actually ended up with this in 5.0?

**Comment by MaximMikhisor** [comment]
Just small remark about this issue.

EF Core 3.1 contains this changes:
https://github.com/dotnet/efcore/commit/5fec0d085ef7e5702208c1169758f4f2d4e996d7#diff-920a9f983bb40645fe12f779d7f2d6df

I guess EF Core 5.0 lost above changes somewhere along the way.

Looking forward to get fix for it. At this moment cannot use 5.0 because of this.

**Comment by r-work** [comment]
Same here, this is blocking from using 5.0.

**Comment by roji** [comment]
I've taken a look, here's what happened:

* The original support for [MaybeNull] was introduced in fb612415df4acdfe36af60f6a183fd925f901deb.
* This commit introduced the `PropertyInfo.GetMethod.ReturnParameter.GetCustomAttribute<MaybeNullAttribute>()` call, which causes a bug on mono (https://github.com/mono/mono/issues/17477). The bug is still unaddressed.
* Commit 5fec0d085ef7e5702208c1169758f4f2d4e996d7 made various changes to make EF Core compatible with netstandard2.0 again. Among these, the generic GetCustomAttribute call was replaced with a string-based attribute lookup method, since the [MaybeNull] attribute does not exist in netstandard2.0 ([here's the change](https://github.com/dotnet/efcore/commit/5fec0d085ef7e5702208c1169758f4f2d4e996d7#diff-a302f86b21919308f13ed7e1f0cdbda7590bcc0bf66db5b378853289d42387a4)).
* This unintentionally also worked around the mono bug, which is apparently triggered by the GetCustomAttribute call. This is why 3.1 doesn't have this bug.
* In 3c5a89a17298d2a762a43d115ae7fca36d46a434, when 3.1 was merged into master (5.0), these changes seem to have been discarded, since master targetted netstandard2.1. This is why the bug exists again in 5.0.
